File: android/app/src/main/java/com/apexedit/editor/core/Time.kt
package com.apexedit.editor.core

import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.roundToLong

/**
 * Frame-accurate time.
 *
 * Editorial maths runs on an integer tick base, never on floats or on
 * microseconds. 705_600_000 ticks per second is chosen so one frame at every
 * supported rate — including the NTSC /1001 rates — and one sample at both
 * 44.1 and 48 kHz land on exact integers:
 *
 *   705600000 / 24            = 29_400_000
 *   705600000 / 25            = 28_224_000
 *   705600000 / 30            = 23_520_000
 *   705600000 / 60            = 11_760_000
 *   705600000 / (30000/1001)  = 23_543_520
 *   705600000 / 48000         = 14_700
 *
 * Media3 speaks microseconds, so [ticksToUs] and [usToTicks] convert at that
 * boundary and nowhere else. Doing editorial arithmetic in microseconds is how
 * cuts end up a frame off at 29.97.
 */
object Time {
    const val TICKS_PER_SECOND: Long = 705_600_000L
    const val US_PER_SECOND: Long = 1_000_000L

    /**
     * 705_600_000 / 1_000_000 is 705.6, so the conversion is scaled by ten to
     * stay in integer arithmetic. At three hours this peaks around 7.6e13 —
     * comfortably inside Long.
     */
    fun ticksToUs(ticks: Long): Long = (ticks * 10L) / 7056L

    fun usToTicks(us: Long): Long = (us * 7056L) / 10L

    fun ticksToMs(ticks: Long): Long = ticksToUs(ticks) / 1000L

    fun msToTicks(ms: Long): Long = usToTicks(ms * 1000L)

    fun secondsToTicks(seconds: Double): Long = (seconds * TICKS_PER_SECOND).roundToLong()

    fun ticksToSeconds(ticks: Long): Double = ticks.toDouble() / TICKS_PER_SECOND
}

/** A frame rate expressed exactly, as a rational. */
data class FrameRate(
    val num: Int,
    val den: Int,
    /** Only meaningful for the /1001 rates; ignored otherwise. */
    val dropFrame: Boolean = false,
) {
    val fps: Double get() = num.toDouble() / den

    /** Exact ticks in one frame. */
    val ticksPerFrame: Long
        get() {
            val exact = Time.TICKS_PER_SECOND * den / num
            require(Time.TICKS_PER_SECOND * den % num == 0L) {
                "Frame rate $num/$den does not divide the tick base exactly"
            }
            return exact
        }

    /** Nominal integer rate used for timecode counting: 30 for 29.97. */
    val nominal: Int get() = Math.round(fps).toInt()

    companion object {
        val FPS_23_976 = FrameRate(24000, 1001)
        val FPS_24 = FrameRate(24, 1)
        val FPS_25 = FrameRate(25, 1)
        val FPS_29_97 = FrameRate(30000, 1001, dropFrame = true)
        val FPS_30 = FrameRate(30, 1)
        val FPS_50 = FrameRate(50, 1)
        val FPS_59_94 = FrameRate(60000, 1001, dropFrame = true)
        val FPS_60 = FrameRate(60, 1)

        val ALL = listOf(
            FPS_23_976, FPS_24, FPS_25, FPS_29_97, FPS_30, FPS_50, FPS_59_94, FPS_60,
        )

        /** Closest supported rate to a measured one, for imported media. */
        fun nearest(fps: Float): FrameRate =
            ALL.minByOrNull { abs(it.fps - fps) } ?: FPS_30
    }
}

fun Long.framesToTicks(rate: FrameRate): Long = this * rate.ticksPerFrame

fun Long.ticksToFrames(rate: FrameRate): Long = floorDiv(rate.ticksPerFrame)

/** Snap down to the frame boundary containing this position. */
fun Long.floorToFrame(rate: FrameRate): Long {
    val tpf = rate.ticksPerFrame
    return floorDiv(tpf) * tpf
}

/** Snap to the nearest frame boundary. */
fun Long.roundToFrame(rate: FrameRate): Long {
    val tpf = rate.ticksPerFrame
    return ((this + tpf / 2).floorDiv(tpf)) * tpf
}

fun Long.ceilToFrame(rate: FrameRate): Long {
    val tpf = rate.ticksPerFrame
    return ((this + tpf - 1).floorDiv(tpf)) * tpf
}

fun Long.isOnFrame(rate: FrameRate): Boolean = this % rate.ticksPerFrame == 0L

data class TimecodeParts(
    val hours: Int,
    val minutes: Int,
    val seconds: Int,
    val frames: Int,
    val negative: Boolean,
)

/**
 * SMPTE timecode with real drop-frame counting: the first two frame numbers of
 * every minute are skipped except on minutes divisible by ten, which is what
 * keeps a 29.97 timeline aligned with wall time.
 */
fun Long.toTimecodeParts(rate: FrameRate): TimecodeParts {
    val negative = this < 0
    val frameIndex = abs(this).ticksToFrames(rate)
    val nominal = rate.nominal

    if (rate.dropFrame && rate.den == 1001) {
        val dropPerMinute = when (nominal) {
            30 -> 2
            60 -> 4
            else -> 0
        }
        if (dropPerMinute > 0) {
            val framesPer10Min = nominal * 60L * 10L - dropPerMinute * 9L
            val framesPerMin = nominal * 60L - dropPerMinute

            val tenMinBlocks = frameIndex / framesPer10Min
            var rem = frameIndex % framesPer10Min

            // The first minute of each ten-minute block drops nothing.
            var minutesInBlock = 0L
            if (rem >= nominal * 60L) {
                rem -= nominal * 60L
                minutesInBlock = 1 + rem / framesPerMin
                rem %= framesPerMin
                rem += dropPerMinute // re-add the skipped frame numbers
            }

            val totalMinutes = tenMinBlocks * 10 + minutesInBlock
            return TimecodeParts(
                hours = (totalMinutes / 60).toInt(),
                minutes = (totalMinutes % 60).toInt(),
                seconds = (rem / nominal).toInt(),
                frames = (rem % nominal).toInt(),
                negative = negative,
            )
        }
    }

    return TimecodeParts(
        hours = (frameIndex / (nominal * 3600L)).toInt(),
        minutes = ((frameIndex / (nominal * 60L)) % 60).toInt(),
        seconds = ((frameIndex / nominal) % 60).toInt(),
        frames = (frameIndex % nominal).toInt(),
        negative = negative,
    )
}

private fun pad2(n: Int) = n.toString().padStart(2, '0')

/** `01:23:45:12`, or `01:23:45;12` for drop-frame. */
fun Long.formatTimecode(rate: FrameRate): String {
    val p = toTimecodeParts(rate)
    val sep = if (rate.dropFrame && rate.den == 1001) ';' else ':'
    val sign = if (p.negative) "-" else ""
    return "$sign${pad2(p.hours)}:${pad2(p.minutes)}:${pad2(p.seconds)}$sep${pad2(p.frames)}"
}

/** Compact `1:23.4` for the timeline ruler, where SMPTE is too dense to read. */
fun Long.formatDuration(tenths: Boolean = false): String {
    val total = Time.ticksToSeconds(this).coerceAtLeast(0.0)
    val h = floor(total / 3600).toInt()
    val m = floor(total / 60).toInt() % 60
    val s = floor(total).toInt() % 60
    val base = if (h > 0) "$h:${pad2(m)}:${pad2(s)}" else "$m:${pad2(s)}"
    return if (tenths) "$base.${floor((total % 1) * 10).toInt()}" else base
}

/** Half-open interval `[start, start + duration)`. */
data class TimeRange(val start: Long, val duration: Long) {
    val end: Long get() = start + duration

    fun contains(t: Long): Boolean = t in start until end

    fun overlaps(other: TimeRange): Boolean = start < other.end && other.start < end

    fun intersect(other: TimeRange): TimeRange? {
        val s = maxOf(start, other.start)
        val e = minOf(end, other.end)
        return if (e > s) TimeRange(s, e - s) else null
    }
}

File: packages/edit-engine/src/time/time.ts
/**
 * Frame-accurate time.
 *
 * Floating-point seconds are not a legal time representation anywhere in this
 * engine. Every timeline position, duration and media offset is an integer
 * number of ticks.
 *
 * TICKS_PER_SECOND is chosen so that a single frame at every frame rate we
 * support is an exact integer number of ticks, including the NTSC /1001 rates
 * and the two common audio sample rates:
 *
 *   705600000 / 24     = 29400000
 *   705600000 / 25     = 28224000
 *   705600000 / 30     = 23520000
 *   705600000 / 50     = 14112000
 *   705600000 / 60     = 11760000
 *   705600000 / (24000/1001) = 29429400
 *   705600000 / (30000/1001) = 23543520
 *   705600000 / (60000/1001) = 11771760
 *   705600000 / 48000  = 14700
 *   705600000 / 44100  = 16000
 *
 * At this rate Number.MAX_SAFE_INTEGER is ~147 days of timeline, which is
 * several orders of magnitude beyond any real project.
 */
export const TICKS_PER_SECOND = 705_600_000;

/** An integer number of ticks. Always exact. */
export type Ticks = number;

/** A frame rate expressed exactly, as a rational. */
export interface FrameRate {
  /** Numerator, e.g. 30000 for 29.97. */
  readonly num: number;
  /** Denominator, e.g. 1001 for 29.97. */
  readonly den: number;
  /**
   * Whether timecode for this rate is displayed drop-frame. Only meaningful
   * for the /1001 rates; ignored otherwise.
   */
  readonly dropFrame?: boolean;
}

export const FRAME_RATES = {
  fps23_976: { num: 24000, den: 1001 } as FrameRate,
  fps24: { num: 24, den: 1 } as FrameRate,
  fps25: { num: 25, den: 1 } as FrameRate,
  fps29_97: { num: 30000, den: 1001, dropFrame: true } as FrameRate,
  fps30: { num: 30, den: 1 } as FrameRate,
  fps50: { num: 50, den: 1 } as FrameRate,
  fps59_94: { num: 60000, den: 1001, dropFrame: true } as FrameRate,
  fps60: { num: 60, den: 1 } as FrameRate,
} as const;

export const ALL_FRAME_RATES: readonly FrameRate[] = Object.values(FRAME_RATES);

/** Exact tick count of one frame at `rate`. Throws if the rate is not exact. */
export function ticksPerFrame(rate: FrameRate): Ticks {
  const exact = (TICKS_PER_SECOND * rate.den) / rate.num;
  if (!Number.isInteger(exact)) {
    throw new RangeError(
      `Frame rate ${rate.num}/${rate.den} does not divide the tick base exactly`,
    );
  }
  return exact;
}

export function frameRateToNumber(rate: FrameRate): number {
  return rate.num / rate.den;
}

export function frameRatesEqual(a: FrameRate, b: FrameRate): boolean {
  return a.num * b.den === b.num * a.den;
}

/**
 * Convert seconds to ticks. Only for crossing the boundary with the outside
 * world (decoders, UI gestures); never for internal arithmetic.
 */
export function secondsToTicks(seconds: number): Ticks {
  return Math.round(seconds * TICKS_PER_SECOND);
}

export function ticksToSeconds(ticks: Ticks): number {
  return ticks / TICKS_PER_SECOND;
}

export function framesToTicks(frames: number, rate: FrameRate): Ticks {
  return Math.round(frames * ticksPerFrame(rate));
}

/** Frame index containing `ticks`. Floors, so a position mid-frame maps to that frame. */
export function ticksToFrames(ticks: Ticks, rate: FrameRate): number {
  return Math.floor(ticks / ticksPerFrame(rate));
}

/** Snap a position down to the frame boundary that contains it. */
export function floorToFrame(ticks: Ticks, rate: FrameRate): Ticks {
  const tpf = ticksPerFrame(rate);
  return Math.floor(ticks / tpf) * tpf;
}

/** Snap a position to the nearest frame boundary. */
export function roundToFrame(ticks: Ticks, rate: FrameRate): Ticks {
  const tpf = ticksPerFrame(rate);
  return Math.round(ticks / tpf) * tpf;
}

export function ceilToFrame(ticks: Ticks, rate: FrameRate): Ticks {
  const tpf = ticksPerFrame(rate);
  return Math.ceil(ticks / tpf) * tpf;
}

/** True when `ticks` sits exactly on a frame boundary. */
export function isOnFrame(ticks: Ticks, rate: FrameRate): boolean {
  return ticks % ticksPerFrame(rate) === 0;
}

/** Snap to an audio sample boundary — used when trimming audio-only clips. */
export function roundToSample(ticks: Ticks, sampleRate: number): Ticks {
  const tps = TICKS_PER_SECOND / sampleRate;
  if (!Number.isInteger(tps)) {
    // Non-exact sample rates fall back to nearest tick, which is sub-microsecond.
    return Math.round(ticks);
  }
  return Math.round(ticks / tps) * tps;
}

export interface TimecodeParts {
  hours: number;
  minutes: number;
  seconds: number;
  frames: number;
  /** True when the value was clamped at zero because the input was negative. */
  negative: boolean;
}

/**
 * SMPTE timecode. Implements true drop-frame counting for the /1001 rates:
 * the first two frame numbers of every minute are skipped except on minutes
 * divisible by ten, which keeps the clock aligned with wall time.
 */
export function ticksToTimecodeParts(ticks: Ticks, rate: FrameRate): TimecodeParts {
  const negative = ticks < 0;
  const frameIndex = Math.max(0, ticksToFrames(Math.abs(ticks), rate));
  const nominal = Math.round(frameRateToNumber(rate)); // 30 for 29.97, 24 for 23.976

  if (rate.dropFrame && rate.den === 1001) {
    const dropPerMinute = nominal === 30 ? 2 : nominal === 60 ? 4 : 0;
    if (dropPerMinute > 0) {
      const framesPer10Min = nominal * 60 * 10 - dropPerMinute * 9;
      const framesPerMin = nominal * 60 - dropPerMinute;

      const tenMinBlocks = Math.floor(frameIndex / framesPer10Min);
      let rem = frameIndex % framesPer10Min;

      // The first minute of each 10-minute block drops nothing.
      let minutesInBlock = 0;
      if (rem >= nominal * 60) {
        rem -= nominal * 60;
        minutesInBlock = 1 + Math.floor(rem / framesPerMin);
        rem = rem % framesPerMin;
        rem += dropPerMinute; // re-add the skipped frame numbers
      }

      const totalMinutes = tenMinBlocks * 10 + minutesInBlock;
      return {
        hours: Math.floor(totalMinutes / 60),
        minutes: totalMinutes % 60,
        seconds: Math.floor(rem / nominal),
        frames: rem % nominal,
        negative,
      };
    }
  }

  return {
    hours: Math.floor(frameIndex / (nominal * 3600)),
    minutes: Math.floor(frameIndex / (nominal * 60)) % 60,
    seconds: Math.floor(frameIndex / nominal) % 60,
    frames: frameIndex % nominal,
    negative,
  };
}

const pad2 = (n: number) => String(n).padStart(2, '0');

/** `01:23:45:12`, or `01:23:45;12` for drop-frame. */
export function formatTimecode(ticks: Ticks, rate: FrameRate): string {
  const p = ticksToTimecodeParts(ticks, rate);
  const sep = rate.dropFrame && rate.den === 1001 ? ';' : ':';
  const sign = p.negative ? '-' : '';
  return `${sign}${pad2(p.hours)}:${pad2(p.minutes)}:${pad2(p.seconds)}${sep}${pad2(p.frames)}`;
}

/** Compact `1:23.4` form for the timeline ruler, where SMPTE is too dense. */
export function formatDuration(ticks: Ticks, opts: { tenths?: boolean } = {}): string {
  const total = Math.max(0, ticksToSeconds(ticks));
  const h = Math.floor(total / 3600);
  const m = Math.floor(total / 60) % 60;
  const s = Math.floor(total) % 60;
  const base = h > 0 ? `${h}:${pad2(m)}:${pad2(s)}` : `${m}:${pad2(s)}`;
  if (!opts.tenths) return base;
  return `${base}.${Math.floor((total % 1) * 10)}`;
}

/** Parse `HH:MM:SS:FF` / `MM:SS:FF` / `SS:FF` timecode back to ticks. */
export function parseTimecode(text: string, rate: FrameRate): Ticks | null {
  const cleaned = text.trim().replace(/;/g, ':');
  if (!/^\d+(:\d+){0,3}$/.test(cleaned)) return null;
  const parts = cleaned.split(':').map((n) => parseInt(n, 10));
  while (parts.length < 4) parts.unshift(0);
  const [h, m, s, f] = parts as [number, number, number, number];
  const nominal = Math.round(frameRateToNumber(rate));
  if (f >= nominal) return null;

  let frameIndex: number;
  if (rate.dropFrame && rate.den === 1001) {
    const dropPerMinute = nominal === 30 ? 2 : nominal === 60 ? 4 : 0;
    const totalMinutes = h * 60 + m;
    const droppedMinutes = totalMinutes - Math.floor(totalMinutes / 10);
    frameIndex = totalMinutes * 60 * nominal + s * nominal + f - droppedMinutes * dropPerMinute;
  } else {
    frameIndex = ((h * 60 + m) * 60 + s) * nominal + f;
  }
  return framesToTicks(frameIndex, rate);
}

export function clampTicks(value: Ticks, min: Ticks, max: Ticks): Ticks {
  return value < min ? min : value > max ? max : value;
}

/** Half-open interval `[start, start + duration)`. */
export interface TimeRange {
  readonly start: Ticks;
  readonly duration: Ticks;
}

export const rangeEnd = (r: TimeRange): Ticks => r.start + r.duration;

export function rangesOverlap(a: TimeRange, b: TimeRange): boolean {
  return a.start < rangeEnd(b) && b.start < rangeEnd(a);
}

export function rangeContains(r: TimeRange, t: Ticks): boolean {
  return t >= r.start && t < rangeEnd(r);
}

export function intersectRanges(a: TimeRange, b: TimeRange): TimeRange | null {
  const start = Math.max(a.start, b.start);
  const end = Math.min(rangeEnd(a), rangeEnd(b));
  return end > start ? { start, duration: end - start } : null;
}
